### ISSUE TITLE
Evaluate Profile-Guided Optimization (PGO)

### ISSUE THREAD

**Issue by zamazan4ik** [wazuh/wazuh#18151]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| any | whole solution | any | any | any |

Hi!

Recently I checked Profile-Guided Optimization (PGO) improvements on multiple projects. The results are [here](https://github.com/zamazan4ik/awesome-pgo/).

Since PGO showed measurable improvements in many workloads I think it could be useful to check PGO on Wazuh.

We need to perform PGO benchmarks on Wazuh. And if it shows improvements - add a note about possible improvements Wazuh performance with PGO. Providing an easier way (e.g. a build option) to build scripts with PGO can be useful for the end-users too.

P.S. Not sure if is it right for such an issue. If you know a better place for the issue (maybe Discussions?) - feel free to convert it to the right type.

**Comment by sempervictus** [comment]
I think that this might be a bit more nuanced than with other projects. Hot paths in the C code will vary to some degree depending on the data and decoder/rule workloads running through the pipeline. IMO, having the code able to take advantage of the various SIMD pipes available on modern hardware would provide a lot more flexibility for the optimizer to choose optimal codepaths; but given the breadth of kit and workloads, there may be a price to pay for over-optimization to a specific set of workloads and HW for folks using the software differently.
Building `-O3` with LTO and cleaning up the code to let the compiler reason better about what it's doing tends to offer "generic benefit to all," but it can get pretty pedantic in terms of the C machinations (take a look at what ZFS has done over the last ~2y).

**Comment by zamazan4ik** [comment]
> I think that this might be a bit more nuanced than with other projects. Hot paths in the C code will vary to some degree depending on the data and decoder/rule workloads running through the pipeline.

Yes, it could be an issue. But if PGO will be integrated as an additional build option, users will be able to optimize Wazuh for their own specific workloads.

> IMO, having the code able to take advantage of the various SIMD pipes available on modern hardware would provide a lot more flexibility for the optimizer to choose optimal codepaths;

Agreed. But integrating PGO does not mean that you cannot use SIMD-like optimizations - you can do all at the same time: LTO, more SIMD-aggressive stuff, PGO (I would say LTO is recommended to do with PGO). Regarding the priorities between these kind of optimizations - well, that's up to you at all since you know about the project much more than me :)

**Comment by sempervictus** [comment]
I wonder how many users would be "down for profiling themselves" for a few weeks to build out a variety of guidance profiles for analysis and homologation... Agree that you "can do it all," just figure a broad dataset is needed to avoid footgun conditions.

**Comment by zamazan4ik** [comment]
> for a few weeks

Honestly, you do not need profiling for weeks to create a PGO-optimized build. Usually, it's enough to collect the required profiles for a much less period of time (minutes/hours). How many users need it? Well, I think it depends on the performance benefits of PGO for Wazuh. If there is a way to collect feedback from the users - would be fine to hear their opinion too.
